clients/ranchercli/client.go

```diff
@@ -0,0 +1,112 @@
+package ranchercli
+
+import (
+	"fmt"
+	osys "os"
+	"os/exec"
+	"strings"
+
+	"github.com/rancher/shepherd/pkg/session"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	rancher   = "./rancher"
+	createCmd = "create"
+	deleteCmd = "delete"
+	moveCmd   = "move"
+)
+
+type Client struct {
+	Session *session.Session
+}
+
+func (c *Client) Create(name string, args ...string) error {
+	args = append([]string{name, createCmd}, args...)
+	c.ExecuteCommand(rancher, args...)
+
+	c.Session.RegisterCleanupFunc(func() error {
+		return c.Delete(name, args...)
+	})
+
+	return nil
+}
+
+func (c *Client) Delete(name string, args ...string) error {
+	args = append([]string{name, deleteCmd}, args...)
+	c.ExecuteCommand(rancher, args...)
+
+	return nil
+}
+
+func (c *Client) Exists(rancher, resourceType, name string) error {
+	cmdStr := fmt.Sprintf("%s %s ls | grep %s", rancher, resourceType, name)
+	cmd := exec.Command("sh", "-c", cmdStr)
+	err := cmd.Run()
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			if exitError.ExitCode() != 0 {
+				return fmt.Errorf("error validating %s exists: %v", name, err)
+			}
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+// ExecuteCommand will execute a command and log any errors.
+func (c *Client) ExecuteCommand(name string, args ...string) error {
+	command := exec.Command(name, args...)
+	_, err := command.Output()
+	if err != nil {
+		return fmt.Errorf("error executing command '%s': %v", name, err)
+	}
+
+	return nil
+}
+
+// NewClient will download the CLI and login to the Rancher server.
+func NewClient(session *session.Session, version, token, host, projectID string) (*Client, error) {
+	client := &Client{
+		Session: session,
+	}
+
+	cmd := exec.Command("sh", "-c", "uname | awk '{print tolower($0)}'")
+	stdout, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	os := strings.TrimSpace(string(stdout))
+
+	logrus.Infof("Setting up the Rancher CLI...")
+	err = client.ExecuteCommand("wget", "https://github.com/rancher/cli/releases/download/"+version+"/rancher-"+os+"-amd64-"+version+".tar.gz")
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.ExecuteCommand("tar", "-xf", "rancher-"+os+"-amd64-"+version+".tar.gz")
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.ExecuteCommand("rm", "rancher-"+os+"-amd64-"+version+".tar.gz")
+	if err != nil {
+		return nil, err
+	}
+
+	err = osys.Chdir("rancher-" + version)
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Infof("Logging into Rancher server...")
+	err = client.ExecuteCommand("./rancher", "login", "--token", token, "https://"+host, "--skip-verify", "--context", projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{}, nil
```

Comment:
Why are you returning an empty struct?

Reply:
That was from a prior iteration, I'll return the client. Good catch!

Reply:
Resolved.